torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3459,15 +3459,11 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                             f"Device mismatch: p={p.device} "  # type: ignore[attr-defined]
                             f"p._saved_grad_shard={p._saved_grad_shard.device}"
                         )
-                        # Check if post-backward was called for this param (FSDP unit).
```

Comment:
@rohan-varma I think this logic can be simplified as shown in the PR. As long as we reshard, then we can forward `_saved_grad_shard` to `.grad`, even if this is the old gradient (i.e. the post-backward hook did not run). This works as long as we reorder the `_catch_all_reshard()` to be before `_finalize_params()`.

Reply:
Also, do you know why we would need to revisit this logic for non-recursive wrapping?

Reply:
Hmm, I seem to remember we had some sort of discussion about it, but not completely sure. 

But why would we want to write an old `_saved_grad_shard` into p.grad? That should have already happened right, the last time the post backward hook did run?